server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java

```diff
@@ -40,7 +47,34 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         mergeRequest.maxNumSegments(request.paramAsInt("max_num_segments", mergeRequest.maxNumSegments()));
         mergeRequest.onlyExpungeDeletes(request.paramAsBoolean("only_expunge_deletes", mergeRequest.onlyExpungeDeletes()));
         mergeRequest.flush(request.paramAsBoolean("flush", mergeRequest.flush()));
-        return channel -> client.admin().indices().forceMerge(mergeRequest, new RestToXContentListener<>(channel));
+        if (request.paramAsBoolean("wait_for_completion", true)) {
+            return channel -> client.admin().indices().forceMerge(mergeRequest, new RestToXContentListener<>(channel));
+        } else {
+            mergeRequest.setShouldStoreResult(true);
+        }
+        /*
+         * Let's try and validate before forking so the user gets some error. The
+         * task can't totally validate until it starts but this is better than
+         * nothing.
+         */
+        ActionRequestValidationException validationException = mergeRequest.validate();
+        if (validationException != null) {
+            throw validationException;
+        }
+        return sendTask(
+            client.getLocalNodeId(),
+            client.executeLocally(ForceMergeAction.INSTANCE, mergeRequest, LoggingTaskListener.instance())
+        );
```

Comment:
nit: shall we move this block in the `else` block so we don't have "leaking" returns?

Reply:
ok, I will move this block in the else block.